### PR TITLE
Graceyi/feature/scheduler

### PR DESCRIPTION
**Pull request by Taylorr82**

### CODE REVIEW

File: src/groundStation/system.py [hunk 1]
```diff
@@ -76,6 +76,61 @@ def __init__(self):
             'LAST': 31
         }
         self.SERVICES = {
+            'SCHEDULER': {
+                'port': 15,  # share a port with EPS time service
```
Comment (Taylorr82) [inline src/groundStation/system.py:80]: I think sharing ports is okay, but you shouldn't mention it. Has a chance of confusing people

File: src/groundStation/system.py [hunk 2]
```diff
@@ -76,6 +76,61 @@ def __init__(self):
...
+                # TODO: these need a error response value
+                'subservice': {
+                    'SET_SCHEDULE': {
+                        'subPort': 0,
+                        'inoutInfo': {
+                            'args': None,  # All scheduled commands should be stored in schedule.txt
```
Comment (Taylorr82) [inline src/groundStation/system.py:86]: This is no longer true after adding the argparse stuff

File: src/groundStation/system.py [hunk 3]
```diff
@@ -76,6 +76,61 @@ def __init__(self):
...
+                            'returns': {
+                                'err': '>b', # Returns 0 on success. Refer for schedule.h for calloc error code. Refer to rederrno.h for reliance edge error codes
+                                'number of cmds scheduled': '>b'    #Returns the number of cmds left in the schedule file. Returns -1 if an error occurred.
```
Comment (Taylorr82) [inline src/groundStation/system.py:89]: There's a 'What' key that allows you to place these comments. They will then get transferred to the autogenerated docs

File: src/cli.py [hunk 4]
```diff
@@ -51,8 +51,8 @@ def cli():
                 data[0] == sysVals.SERVICES.get('SCHEDULER').get('subservice').get('DELETE_SCHEDULE').get('subPort') or
                 data[0] == sysVals.SERVICES.get('SCHEDULER').get('subservice').get('REPLACE_SCHEDULE').get('subPort'))
                 ):
-
-                embeddedCSPObj = groundStation.getEmbededCSPData(data)
+                filename = opts.getOptions()
```
Comment (Taylorr82) [inline src/cli.py:54]: This isn't quite how the options class works

File: src/groundStation/groundStation.py [hunk 5]
```diff
@@ -397,15 +397,22 @@ def getOptions(self):
             type=str,
             default="EX2",
             help='Satellite parameter for automatic programs (e.g FTP)')
+
+        self.parser.add_argument(
```
Comment (Taylorr82) [inline src/groundStation/groundStation.py:401]: Should have a more descriptive name than just --file

File: src/cli.py (hunk 4 above)
Comment (yunqingyi) [reply src/cli.py:54]: yeah this was my experiment...it should be correct now

File: src/cli.py [hunk 6]
```diff
@@ -39,18 +42,32 @@ def cli():
...
                 print("calling __setPIPE")
                 gs.__setPIPE__()
+                data = bytearray(libcsp.packet_get_data(toSend))
+
+            if (
+                server == sysVals.APP_DICT.get('OBC') and 
+                port == sysVals.SERVICES.get('SCHEDULER').get('port') and
+                (data[0] == sysVals.SERVICES.get('SCHEDULER').get('subservice').get('SET_SCHEDULE').get('subPort') or
+                data[0] == sysVals.SERVICES.get('SCHEDULER').get('subservice').get('DELETE_SCHEDULE').get('subPort') or
+                data[0] == sysVals.SERVICES.get('SCHEDULER').get('subservice').get('REPLACE_SCHEDULE').get('subPort'))
```
Comment (jmmabanta) [inline src/cli.py:52]: How would 'REPLACE_SCHEDULE' be handled differently compared to 'SET_SCHEDULE' (Does the scheduler file need to look different?) , and how would 'DELETE_SCHEDULE' be handled? All these three seem to follow the same process of reading a schedule file so I'm just wondering if I need to do something special

File: src/cli.py [hunk 7]
```diff
@@ -39,18 +42,32 @@ def cli():
                 # Can be deleted for flight
                 print("calling __setPIPE")
                 gs.__setPIPE__()
+                data = bytearray(libcsp.packet_get_data(toSend))
+
+            if (
+                server == sysVals.APP_DICT.get('OBC') and
```
Comment (jmmabanta) [inline src/cli.py:48]: I believe this was mentioned in the previous PR but the server prefixes are now 'EX2', 'ARI', and 'YUK'

File: src/cli.py [hunk 8]
```diff
@@ -26,9 +26,12 @@
 flag = groundStation.GracefulExiter()
 
 def cli():
+
+    sysVals = groundStation.SystemValues()
+
     while True:
         if flag.exit():
-            print('Exiting receiving loop')
+            print('Exiting receiving loop\n')
```
Comment (rcunrau) [inline src/cli.py:34]: Should we reduce the amount of debug output as this gets more stable?

**Review by rcunrau — DISMISSED** [review]

File: src/groundStation/groundStation.py [hunk 9]
```diff
@@ -400,7 +404,7 @@ def getOptions(self):
 
         self.parser.add_argument(
             '-f',
```
Comment (Taylorr82) [inline src/groundStation/groundStation.py:406]: -f is used by the updater service which derives this class. This should have a different single character option, or just none at all

File: src/groundStation/system.py (hunk 3 above)
Comment (Taylorr82) [reply src/groundStation/system.py:89]: @yunqingyi

File: src/groundStation/system.py (hunk 2 above)
Comment (Taylorr82) [reply src/groundStation/system.py:86]: @yunqingyi

File: src/groundStation/groundStation.py (hunk 9 above)
Comment (Taylorr82) [reply src/groundStation/groundStation.py:406]: @yunqingyi

File: src/cli.py (hunk 6 above)
Comment (Taylorr82) [reply src/cli.py:55]: @yunqingyi

File: src/groundStation/groundStation.py (hunk 5 above)
Comment (Taylorr82) [reply src/groundStation/groundStation.py:405]: This change hasn't made it to github

File: src/groundStation/system.py (hunk 2 above)
Comment (yunqingyi) [reply src/groundStation/system.py:86]: the argparse stuff only selects which text file to upload, it doesn't get passed as a parameter to the subservice

File: src/groundStation/groundStation.py [hunk 10]
```diff
@@ -399,7 +399,7 @@ def getOptions(self):
             help='Satellite parameter for automatic programs (e.g FTP)')
 
         self.parser.add_argument(
-            '-f',
+            '-sched',
```
Comment (Taylorr82) [inline src/groundStation/groundStation.py:402]: It's not typical for single dash options to be longer than 1 character. Maybe the short form should just be removed

File: src/groundStation/system.py (hunk 1 above)
Comment (yunqingyi) [reply src/groundStation/system.py]: pushed

File: src/groundStation/system.py (hunk 3 above)
Comment (yunqingyi) [reply src/groundStation/system.py]: fixed

File: src/groundStation/groundStation.py (hunk 5 above)
Comment (yunqingyi) [reply src/groundStation/groundStation.py:401]: -S used instead

File: src/cli.py (hunk 6 above)
Comment (yunqingyi) [reply src/cli.py:55]: REPLACE_SCHEDULE deletes the existing one, then starts a new schedule. SET_SCHEDULE adds onto the existing schedule

**Comment by Taylorr82** [comment]
I think this is ready to go, but i can't approve it myself because the PR technically belongs to me

**Review by rcunrau — APPROVED** [review]

**Review by jmmabanta — APPROVED** [review]

**Pull request merged**